### ISSUE TITLE
Relax attribute requirements when registering ArtifactTransforms

### ISSUE THREAD

**Issue by tresat** [gradle/gradle#30786]
Followup to https://github.com/gradle/gradle/pull/30768.

We should relax the restrictions on registering ArtifactTransforms where every "to" attribute must have a corresponding "from" attribute.  A transform that takes checks every variant with "type=text" and outputs an AI summary of the text with "summarized=true" for instance.

We should also remove the requirement for any "from" attributes whatsoever.  A transform that only adds attributes is sensible: for instance it could validate the artifact sets for some prohibited content or size and add an "approved=true" attribute to any variant.

**Comment by jvandort** [comment]
This is a potential solution to help push https://github.com/gradle/gradle/issues/31538 forward

**Comment by big-guy** [comment]
I don't think we should special case this. Wildcard matching is dangerously close to compatibility rules and makes transforms "sticky" in bad ways. We don't really deal with unset values in a systematic way, so a compatibility rule can't tell Gradle how to treat the lack of a value or a query can't be written to match a lack of value. I think that's what's missing here.

The restriction is that all `to` attributes are represented in `from`, which means that you can be more specific (more `from` attributes) when selecting a source variant, but you can't add arbitrary attributes (extra `to` attributes). 

What are the scenarios where you would have a `to` attribute, but not a `from`?
1. The source variant _does_ have the extra attribute, but you don't care about its value and you do want to match and replace it.
2. The source variant _does_ have the extra attribute and you _do not_ want to match.
3. The source variant _does not_ have the extra attribute and do you want to match and replace it.
4. The source variant _does not_ have the extra attribute and you _do not_ want to match.

The lack of a `from` value for an attribute on the `to` side could be interpreted in a few ways:
- Match any value, including unset. This is what (1) and (3) want.
- Match any value, excluding unset. This is what (4) wants. 
- Match unset. This is what (2) and (3) want. 

I think the scenario Tom has in mind is (3), so if a transform could be registered with something like:
```
from(/* other attributes */)
from(Approved, UNSET) 
to(Approved, "true")
```
...that covers the useful scenarios without opening up the ambiguity of the other scenarios. 

We would still need to be careful that you're able to register transforms with enough specificity, so maybe we'd need to require that all to attributes are included in from _and_ from does not consist of only unset attributes. I think we want to avoid this being a valid registration:
```
// no from attributes
to(Approved, "true")
```

I would want to see a good use case, but if you did want to register a transform that said any value of attribute X is allowed and it always transforms it into value AAA, you can accomplish this with a transform that requests attribute X with value ZZZ and a compatibility rule that said ZZZ is compatible with anything.
